### PR TITLE
direct_mapping: fix iter_memory_pair_chunks in reverse mode

### PR DESCRIPTION
**Pull request by alessandrod**
iter_memory_pair_chunks was iterating regions in reverse, but not memory _within_ regions in reverse, which broke some overlapping memmoves.

This commit fixes the issue and simplifies the implementation by removing nested loops which made control flow hard to reason about.

I've added (a lot of) tests that compare libc's memmove against our memmove.

### CODE REVIEW

**Review by BoredPerson — APPROVED** [review]
Looks good to me! There are still some weird edge cases with copies going over the gaps in the stack but that's a more general issue with the gap implementation and not this code.

**Comment by alessandrod** [comment]
> Looks good to me! There are still some weird edge cases with copies going over the gaps in the stack but that's a more general issue with the gap implementation and not this code.

Thanks for reviewing! Yes we're aware of those, the plan is to disable gaps when we enable direct mapping.

**Comment by stegaBOB** [comment]
I can't really talk on the actual code changes here, but just chiming in to confirm that this does resolve the memmove issue we were running into with direct mapping on 1.16.x and all existing 1.17.x versions. Here's the minimal example program we used to test it: https://beta.solpg.io/65626e1afb53fa325bfd0c34. The "Can do overlapping memmove" test was failing before this PR.

File: programs/bpf_loader/src/syscalls/mem_ops.rs [hunk 1]
```diff
@@ -734,72 +774,59 @@ mod tests {
...
+    #[test_case(&[2, 9], (3, 0, 5) ; "two regions partial overlap start")]
+    #[test_case(&[3, 9], (1, 2, 5) ; "two regions partial overlap middle")]
+    #[test_case(&[7, 3], (2, 6, 4) ; "two regions partial overlap end")]
+    #[test_case(&[2, 6, 3, 4], (0, 10, 2) ; "many regions no overlap, single source region")]
+    #[test_case(&[2, 1, 2, 5, 6], (2, 10, 4) ; "many regions no overlap, multiple source regions")]
+    #[test_case(&[8, 1, 3, 6], (0, 0, 18) ; "many regions complete overlap")]
+    #[test_case(&[7, 3, 1, 4, 5], (5, 0, 8) ; "many regions overlap start")]
+    #[test_case(&[1, 5, 2, 9, 3], (5, 4, 8) ; "many regions overlap middle")]
+    #[test_case(&[3, 9, 1, 1, 2, 1], (2, 9, 8) ; "many regions overlap end")]
+    fn test_memmove_non_contiguous(
```
Comment (Lichtso) [inline programs/bpf_loader/src/syscalls/mem_ops.rs:796]: Why is the test called non-contiguous?
From what I can tell `build_memory_mapping()` can only generate a contiguous memory mapping, without any gaps.

Comment (alessandrod) [reply programs/bpf_loader/src/syscalls/mem_ops.rs:796]: Because it tests `memmove_non_contiguous`. Non contiguous as in the underlying allocations are not linear (many mallocs, many regions).
